File: dart/binary-search-tree/lib/binary_search_tree.dart
class BinarySearchTree {
  late final TreeNode root;

  List<String> get sortedData => _calcSortedData();

  BinarySearchTree(String rootValue) {
    root = TreeNode(rootValue);
  }

  void insert(String value) => root.append(value);

  List<String> _calcSortedData() => root.sort();
}

class TreeNode {
  final String data;

  TreeNode? left;
  TreeNode? right;

  TreeNode(this.data);

  void append(String value) {
    final v = int.parse(value);

    if (v <= int.parse(data)) {
      if (left == null) {
        left = TreeNode(value);
      } else {
        left?.append(value);
      }
    } else {
      if (right == null) {
        right = TreeNode(value);
      } else {
        right?.append(value);
      }
    }
  }

  List<String> sort() => [...left?.sort() ?? [], data, ...right?.sort() ?? []];
}

File: kotlin/binary-search-tree/src/main/kotlin/BinarySearchTree.kt
class BinarySearchTree<T : Comparable<T>> {

    data class Node<T>(val data: T, var left: Node<T>? = null, var right: Node<T>? = null)

    var root: Node<T>? = null

    fun insert(value: T) {
        if (root == null) {
            root = Node(value)
            return
        }

        var node = root

        while (true) {
            if (value <= node!!.data) {
                if (node.left == null) {
                    node.left = Node(value)
                    return
                }
                node = node.left
            } else {
                if (node.right == null) {
                    node.right = Node(value)
                    return
                }
                node = node.right
            }
        }

    }

    fun asSortedList(): List<T> {
        val result = mutableListOf<T>()

        fun inorder(node: Node<T>?) {
            if (node == null) return
            inorder(node.left)
            result.add(node.data)
            inorder(node.right)
        }

        inorder(root)
        return result
    }

    fun asLevelOrderList(): List<T> {
        val result = mutableListOf<T>()
        val queue = ArrayDeque<Node<T>>()

        root?.let { queue.add(it) }

        while (queue.isNotEmpty()) {
            val node = queue.removeFirst()
            result.add(node.data)

            node.left?.let { queue.add(it) }
            node.right?.let { queue.add(it) }
        }

        return result
    }

}
